chips/segger/src/rtt.rs

```diff
@@ -278,11 +292,36 @@ impl<'a, A: hil::time::Alarm<'a>> uart::Transmit<'a> for SeggerRtt<'a, A> {
 
 impl<'a, A: hil::time::Alarm<'a>> hil::time::AlarmClient for SeggerRtt<'a, A> {
     fn alarm(&self) {
-        self.client.map(|client| {
-            self.client_buffer.take().map(|buffer| {
+        self.tx_client.map(|client| {
+            self.tx_client_buffer.take().map(|buffer| {
                 client.transmitted_buffer(buffer, self.tx_len.get(), Ok(()));
             });
         });
+        self.rx_client.map(|client| {
+            self.rx_client_buffer.take().map(|buffer| {
+                self.config.map(|config| {
+                    let write_position = &config.down_buffer.write_position;
+                    let read_position = &config.down_buffer.read_position;
+                    while self.rx_cursor.get() < self.rx_len.get()
+                        && write_position.get() != read_position.get()
+                    {
+                        buffer[self.rx_cursor.get()] =
+                            config.down_buffer[read_position.get() as usize].get();
+                        self.rx_cursor.set(self.rx_cursor.get() + 1);
+                        read_position.set((read_position.get() + 1) % config.down_buffer.length);
+                        fence(Ordering::SeqCst);
```

Comment:
Just trying to reason about why this `fence` is required in addition to the one below: I'm assuming that it's to protect against compiler reorderings where the `read_position` is updated before the buffer contents were actually read?

In this case, I think the `fence` should be moved in between the `down_buffer` read, and updating of the cursor / read position.

Then the last `fence` after this loop takes care of flushing the final write.

Also, I think this should have a `fence` before the initial `write_position.get()` in the while-loop conditional.